Many test failures on aarch64

## Expected Behavior

The test suite should pass the same as on x86_64

## Actual Behavior
Multiple test fail (none of which fail on x86_64 with the same software environment except for the Linux kernel)
```
starting phase `check'
make: Entering directory '/tmp/guix-build-proot-5.2.0-alpha.drv-0/source/test'
which: no valgrind in (/gnu/store/k6lhgdbxkix3fk1bw71bymsyv0ibi411-which-2.21/bin:/gnu/store/wnssqwknkb5kdprn47zf2m0s7faar57q-util-linux-2.35.1/bin:/gnu/store/wnssqwknkb5kdprn47zf2m0s7faar57q-util-linux-2.35.1/sbin:/gnu/store/a60cldyjgqjcxnaxgi7rmk2sgsnbf3dv-coreutils-minimal-8.30/bin:/gnu/store/hjqd2irmhgdb0kcbxaciar8vkkhkb62i-pkg-config-0.29.2/bin:/gnu/store/15byiv88gdrpgpjpzk2pdz5pj546q0ir-python-docutils-0.16/bin:/gnu/store/91sinv421wls6miwwyrj15fh865jgmj3-libarchive-3.4.2/bin:/gnu/store/ggfdxvdgbrjvg2cx3mm7yv5j1y433b01-tar-1.32/bin:/gnu/store/ink6zpk9adb2k2w6a0qif3d45vazl7dv-gzip-1.10/bin:/gnu/store/n8dfjyw11airbnfzz192p9bdbgf31kvj-bzip2-1.0.8/bin:/gnu/store/ggq21is2vyq1jkkqrs90brmwayc4rnmy-xz-5.2.4/bin:/gnu/store/qlxznrl2zzivcln9zgyzw2zshqkjirfs-file-5.38/bin:/gnu/store/yl67anl7279qzrd3kxzj97541jzr2k3d-diffutils-3.7/bin:/gnu/store/w5n1p8kj9zn4zmjhr0bjbknakxij67w2-patch-2.7.6/bin:/gnu/store/x03dmf3xhi3il3mx383zpl4zic38iz1p-findutils-4.7.0/bin:/gnu/store/j80hcnd9p1hh0yb5j1r0gn4vd329a78n-gawk-5.0.1/bin:/gnu/store/wk5vdnms96n2iqfpqyg9gl5lhg8vdc32-sed-4.8/bin:/gnu/store/f3qp8l7f6kmfrvwnyix35kax2lv7lmnb-grep-3.4/bin:/gnu/store/wiv74mq31s0vppzm99qwnjkd8xa3albw-coreutils-8.32/bin:/gnu/store/j7fqszmnydxix8l35d7l3b7dj44qjrw4-make-4.3/bin:/gnu/store/7sfbiqh21h90bc6zi8br1xh60m6qgdd5-bash-minimal-5.0.16/bin:/gnu/store/5m8wvcahy3dgpyhxlq5aq85zbvh5jqfz-ld-wrapper-0/bin:/gnu/store/5371r5kyj50l6176lg6jbfzgfgb2xh14-binutils-2.34/bin:/gnu/store/ap7hgyv4rjqmhg4a6cb6cypsh3g1f5q4-gcc-7.5.0/bin:/gnu/store/cb88z63hyg1icd2kkahiink2p291mhr2-glibc-2.31/bin:/gnu/store/cb88z63hyg1icd2kkahiink2p291mhr2-glibc-2.31/sbin:/bin)
  CHECK test-16573e73 FAILED
  CHECK test-1ffc8309 FAILED
  CHECK test-25069c12 FAILED
  CHECK test-25069c13 FAILED
  CHECK test-16573e73 FAILED
  CHECK test-5bed7141 FAILED
  CHECK test-5bed7143 FAILED
  CHECK test-82ba4ba1 FAILED
  CHECK test-9c07fad8 FAILED
  CHECK test-5bed7143 FAILED
  CHECK test-82ba4ba1 ok
  CHECK test-bdc90417 FAILED
  CHECK test-c5a7a0f0 FAILED
  CHECK test-fa205b56 FAILED
  CHECK test-fdf487a0 FAILED
  CHECK test-c5a7a0f0 ok
  CHECK test-iiiiiiii FAILED
  CHECK test-00000000 ok
  CHECK test-0238c7f1 ok
  CHECK test-0830d8a8 skipped
  CHECK test-071599da FAILED
  CHECK test-03969e70 ok
  CHECK test-092c5e26 FAILED
  CHECK test-1743dd3d ok
  CHECK test-1cd9d8f9 ok
  CHECK test-1ffc8309 ok
  CHECK test-22222222 ok
  CHECK test-230f47ch skipped
  CHECK test-2401b850 FAILED
  CHECK test-1fedd9a3 ok
  CHECK test-305ae31d ok
  CHECK test-311b7a95 FAILED
  CHECK test-3624be91 FAILED
  CHECK test-3dec4597 ok
  CHECK test-517e1d6b ok
  CHECK test-517e1d6a FAILED
  CHECK test-53355a5b ok
  CHECK test-55b731d3 ok
  CHECK test-5467b986 FAILED
  CHECK test-5996858d skipped
  CHECK test-55fd1da5 ok
  CHECK test-5bed7142 ok
  CHECK test-67972fbe skipped
  CHECK test-691786c8 skipped
  CHECK test-654decce FAILED
  CHECK test-6d1e2650 ok
  CHECK test-6b5a254a ok
  CHECK test-6fb08ce1 ok
  CHECK test-7601199b FAILED
  CHECK test-82ba4ba1 skipped
  CHECK test-8a83376a skipped
  CHECK test-713b6910 FAILED
  CHECK test-99999999 FAILED
  CHECK test-9f5eeb72 skipped
  CHECK test-8e5fa256 ok
  CHECK test-a4d7ed70 ok
  CHECK test-b161bc0a ok
  CHECK test-aaaaaaaa FAILED
  CHECK test-b94dd86a ok
  CHECK test-b6df3cbe FAILED
  CHECK test-c15999f9 ok
  CHECK test-c68d18dc skipped
  CHECK test-bbbbbbbb ok
  CHECK test-11111111 FAILED
  CHECK test-careauth skipped
  CHECK test-careexit skipped
  CHECK test-carehwcp skipped
  CHECK test-carequot skipped
  CHECK test-c6b77b77 ok
  CHECK test-cccccccc ok
  CHECK test-cea75343 ok
  CHECK test-chroot01 ok
  CHECK test-cb1143ab ok
  CHECK test-commmmmm ok
  CHECK test-d1da0d8d ok
  CHECK test-d1be631a ok
  CHECK test-d2175fc3 ok
  CHECK test-d92b57ca ok
  CHECK test-de756935 ok
  CHECK test-df4de4db skipped
  CHECK test-dfb0c3b6 FAILED
  CHECK test-e87ca6ca skipped
  CHECK test-dddddddd ok
  CHECK test-e99993c8 ok
  CHECK test-eddeba0e ok
  CHECK test-e940896f ok
  CHECK test-f7089d4f skipped
  CHECK test-fbca9cc2 skipped
  CHECK test-getres32 skipped
  CHECK test-ffffffff ok
  CHECK test-getresid ok
  CHECK test-gggggggg ok
  CHECK test-hhhhhhhh skipped
  CHECK test-killexit FAILED
  CHECK test-mmmmmmmm ok
  CHECK test-proocare skipped
  CHECK test-pppppppp ok
  CHECK test-rrrrrrrr ok
  CHECK test-python01 ok
  CHECK test-wwwwwwww FAILED
  CHECK test-tempdire ok
  CHECK test-0cf405b0 FAILED
  CHECK test-1c68c218 FAILED
  CHECK test-305ae31d FAILED
  CHECK test-33333333 FAILED
  CHECK test-33333334 FAILED
  CHECK test-44444444 FAILED
  CHECK test-51943658 FAILED
  CHECK test-yyyyyyyy ok
  CHECK test-66666666 FAILED
  CHECK test-79cf6614 FAILED
  CHECK test-88888888 FAILED
  CHECK test-a8e69d6f FAILED
  CHECK test-af062114 FAILED
  CHECK test-bug-138 FAILED
  CHECK test-c10e2073 FAILED
  CHECK test-d2175fc4 FAILED
  CHECK test-e87b34ae FAILED
  CHECK test-nnnnnnnn FAILED
  CHECK test-oooooooo FAILED
  CHECK test-ptrace00 FAILED
  CHECK test-xxxxxxxx FAILED
  CHECK test-cdd39012 ok
make: *** [GNUmakefile:22: check_failure] Error 1
make: Leaving directory '/tmp/guix-build-proot-5.2.0-alpha.drv-0/source/test'
command "make" "check" "-C" "test" "-j" "3" failed with status 2
```

## Steps to Reproduce the Problem

  1. Run the test suite with: `make check -C test V=1 -j 4`.

## Specifications
```
  - Proot/Care version: v5.2.0-alpha tag
  - Kernel version: 5.8.13-gnu #1 SMP 1 aarch64 GNU/Linux
  - Host distribution: Guix System
  - Guest distribution: N/A
```
Using the following build inputs:
```
coreutils-minimal@8.30
libarchive@3.4.2
make@4.2.1
pkg-config@0.29.2
talloc@2.3.2
util-linux@2.35.1
which@2.21
```

## Detailed build log
[7254bmg1pr8n7dxqn1qndsyfs2d7lc-proot-5.2.0-alpha.drv.txt](https://github.com/proot-me/proot/files/6788760/7254bmg1pr8n7dxqn1qndsyfs2d7lc-proot-5.2.0-alpha.drv.txt)

--- a/src/arch.h
+++ b/src/arch.h
@@ -33,6 +33,7 @@ typedef unsigned char byte_t;
 
 #define SYSCALL_AVOIDER ((word_t) -2)
 #define SYSTRAP_NUM SYSARG_NUM
+#define STACK_ALIGNMENT 16
 
 #if !defined(ARCH_X86_64) && !defined(ARCH_ARM_EABI) && !defined(ARCH_X86) && !defined(ARCH_SH4)
 #    if defined(__x86_64__)

--- a/src/execve/exit.c
+++ b/src/execve/exit.c
@@ -223,10 +223,9 @@ static int transfer_load_script(Tracee *tracee)
 			: strlen(tracee->load_info->raw_path) + 1);
 
 	/* A padding will be appended at the end of the load script
-	 * (a.k.a "strings area") to ensure this latter is aligned to
-	 * a word boundary, for sake of performance.  */
+	 * (a.k.a "strings area") to ensure this latter is aligned properly. */
 	padding_size = (stack_pointer - string1_size - string2_size - string3_size)
-			% sizeof_word(tracee);
+			% STACK_ALIGNMENT;
 
 	strings_size = string1_size + string2_size + string3_size + padding_size;
 	string1_address = stack_pointer - strings_size;

--- a/src/extension/kompat/kompat.c
+++ b/src/extension/kompat/kompat.c
@@ -605,6 +605,7 @@ static void adjust_elf_auxv(Tracee *tracee, Config *config)
 	word_t stack_pointer;
 	void *argv_envp;
 	size_t size;
+	size_t reserve_size;
 	int status;
 
 	vectors_address = get_elf_aux_vectors_address(tracee);
@@ -671,8 +672,11 @@ static void adjust_elf_auxv(Tracee *tracee, Config *config)
 
 	/* Allocate enough room in tracee's stack for the new ELF
 	 * auxiliary vector.  */
-	stack_pointer   -= 2 * sizeof_word(tracee);
-	vectors_address -= 2 * sizeof_word(tracee);
+	reserve_size = 2 * sizeof_word(tracee);
+	/* Make sure the stack is still aligned */
+	reserve_size = ((reserve_size - 1) / STACK_ALIGNMENT + 1) * STACK_ALIGNMENT;
+	stack_pointer   -= reserve_size;
+	vectors_address -= reserve_size;
 
 	/* Note that it is safe to update the stack pointer manually
 	 * since we are in execve sysexit.  However it should be done

--- a/src/tracee/mem.c
+++ b/src/tracee/mem.c
@@ -508,6 +508,9 @@ word_t alloc_mem(Tracee *tracee, ssize_t size)
 	if (stack_pointer == peek_reg(tracee, ORIGINAL, STACK_POINTER))
 		size += RED_ZONE_SIZE;
 
+	/* Align the stack */
+	size = ((size - 1) / STACK_ALIGNMENT + 1) * STACK_ALIGNMENT;
+
 	/* Sanity check. */
 	if (   (size > 0 && stack_pointer <= (word_t) size)
 	    || (size < 0 && stack_pointer >= ULONG_MAX + size)) {

